src/main/java/com/damnhandy/uri/template/Expression.java

```diff
@@ -350,57 +361,101 @@ else if (varname.lastIndexOf(Modifier.EXPLODE.getValue()) > 0)
+
+            results.put(key, values);
         }
 
-        String unreserved = "[\\w-\\d.~]";
-        String reserved = "[:\\/?#\\[\\]@!$&'()*+;=]"; //todo removing , for now
-        String encoded = "%[A-Fa-f\\d]{2}";
+        Map<String, Either<String, List<String>>> ret = new HashMap<>();
+        results.forEach((k,v) -> ret.put(k, v.size() == 1 ? left(v.get(0)) : right(v)));
 
+        return ret;
+    }
 
-        for(VarSpec v : getVarSpecs()){
+    private Map<String, Either<String, List<String>>> matchSegments(String part){
+        final String prefix = getOperator().getPrefix();
+        final String separator = getOperator().getSeparator();
+        List<String> varNames = getVarSpecs().stream()
+        .map(VarSpec::getVariableName)
+        .collect(toList());
 
-            if(getOperator() == Operator.MATRIX
-            || getOperator() == Operator.CONTINUATION
-            || getOperator() == Operator.QUERY)
-                b.append(v.getVariableName())
-                .append('=');
+        StringBuilder regex = prefix == null || prefix.isEmpty() ? new StringBuilder() : new StringBuilder('\\').append(prefix);
+        Map<String, Either<String, List<String>>> results = new HashMap<>();
 
-            b
-            .append("(?<")
-            .append(v.getVariableName())
-            .append('>');
+        for(String varName:varNames){
+            regex.append("(?<").append(Pattern.quote(varName)).append(">[^").append(separator).append("]*)").append(separator);
+        }
 
-            if(getOperator().getEncoding() == UriTemplate.Encoding.U)
-                b.append("(?:" + unreserved + "|" + encoded + ')');
-            else
-                b
-                .append("(?:")
-                .append(unreserved)
-                .append('|')
-                .append(reserved)
-                .append('|')
-                .append(encoded)
-                .append(')');
-            b.append('*'); //todo this needs to look at the varspec to check for prefix modifier
-            b
-            .append(')')
-            .append(getOperator().getSeparator())
-            .append('?');
+        regex.deleteCharAt(regex.length() -1);
+
+        Pattern pattern = Pattern.compile(regex.toString());
+        Matcher matcher = pattern.matcher(part);
+
+        if (!matcher.matches())
+            throw new RuntimeException("doesn't match");
+
+        for(String varName:varNames){
+            results.put(varName, left(matcher.group(varName)));
         }
 
+        return results;
+    }
+
+    public String getGroupName(){
+        return this.groupName;
+    }
+
+    // todo improve this
+    private Pattern buildMatchingPattern()
+    {
+        String prefix = getOperator().getPrefix();
+        prefix = prefix.equals("+") ? "" : prefix;
+        StringBuilder b = new StringBuilder("(?<")
+        .append(groupName)
+        .append('>')
+        .append('\\')
+        .append(prefix)
+        .append(".{1,9001}")
```

Comment:
Any reason for limiting this to 9001 characters?  You can use `.+` if you want to match at least one character.